### ISSUE TITLE
Cannot parse URI with query/fragment/path containing square brackets ([])

### ISSUE THREAD

**Issue by CIAvash** [raku-community-modules/URI#47]
Although it works in the browser.

```raku
URI.new: 'https://httpbin.org/get?query[0]=test'
```
```
Could not parse URI: https://httpbin.org/get?query[0]=test
  in method parse at ... (URI) line 283
  in method new at ... (URI) line 381
```

**Comment by JJ** [comment]
I'm checking [the grammar](https://www.rfc-editor.org/rfc/rfc3986#page-4) and I don't think square brackets are part of the `pchar` token that defines that part of the query. square brackets are `gen-delim`, (and it's defined as such at the grammar, which is a pretty loyal rendering of the original grammar), but not `sub-delim` which is what's in this OP.
It might be a bit more informative, though, although I don't see how. So I'm for closing this issue, if no one else objects.

**Comment by jonathanstowe** [comment]
Yeah, I think the (quite hard to follow, because it is spread out so much,) grammar in the RFC precludes the `[` and `]` there because, as you say, they aren't in `sub-delim`.  They probably should be "percent encoded" as (e.g.) `%5B' and `%5D`.

It's entirely possible that either the browser is encoding under the hood, or that everything is being more tolerant than the grammar in the RFC.  https://httpbin.org/get?query%5B0%5D=test works fine everywhere.

**Comment by JJ** [comment]
> It's entirely possible that either the browser is encoding under the hood, or that everything is being more tolerant than the grammar in the RFC. https://httpbin.org/get?query%5B0%5D=test works fine everywhere.

Browsers are notable for that kind of thing. You can put whatever you want in the URL slot, it will be encoded.

So... closing?

**Comment by jonathanstowe** [comment]
Yeah I think so. `curl` goes so far as to use the `[]` for its own use (glob the url,) and : 

```
       -g, --globoff
              This  option switches off the "URL globbing parser". When you set this option, you can specify URLs that contain the letters {}[] without having curl itself interpret them. Note that these letters are not normal legal
              URL contents but they should be encoded according to the URI standard.
```
